docs/markdown/Codegen-module.md

```diff
@@ -0,0 +1,101 @@
+- `header`: The optional output name for a header file. If this is unset no header is added
+- `table`: The optional output name for a table file. If this is unset no table will be generated
+                         'POSIX specification does not require that header output be cont')
+The outputs will be in the form `source [header] [table]`, which means those can be accessed by indexing the output of the `lex` call:
+
+```meson
+l1 = codegen.lex('lexer.l', header : '@BASENAME@.h', table : '@BASENAME@.tab.h')
+headers = [l1[1], l1[2]]  # [header, table]
+
+l2 = codegen.lex('lexer.l', table : '@BASENAME@.tab.h')
+table = l2[1]
+```
+
+### find_yacc()
+
+```meson
+codegen.find_yacc(implementations : ['bison', 'win_bison'])
+```
+
+This function provides fine grained controls over which implementation(s) and version(s) of the parser generator to use.
+
+Accepts the following keyword arguments:
+
+ - `implementations`: a string array of acceptable implementations to use. May include: `yacc`, `byacc`, or `bison`.
+ - `yacc_version`: a string array of version constraints to apply to the `yacc` binary
+ - `byacc_version`: a string array of version constraints to apply to the `byacc` binary
+ - `bison_version`: a string array of version constraints to apply to the `bison` binary
+ - `win_bison_version`: a string array of version constraints to apply to the `win_bison` binary
+
+### yacc()
+
+```meson
+codegen.yacc('parser.y')
+```
+
+This function wraps bison, yacc, byacc, and win_bison (on Windows), and attempts to abstract away the differences between them
+
+This requires an input file, which may be a string, File, or generated source. It additionally takes the following options keyword arguments:
+
+- `version`: Version constraints on the lexer
+- `args`: An array of extra arguments to pass the lexer
+- `plainname`: If set to true then `@PLAINNAME@` will be used as the source base, otherwise `@BASENAME@`.
+- `source`: the name of the source output. If this is unset Meson will use `{base}.{ext}` with an extension of `cpp` if the input has an extension of `.yy` or `c` otherwise, with base being determined by the `plainname` argument.
+- `header`: the name of the source output. If this is unset Meson will use `{base}.{ext}` with an extension of `hpp` if the input has an extension of `.yy` or `h` otherwise, with base being determined by the `plainname` argument.
+- `locations`: The name of the locations file, if one is generated. Due to the way yacc works this must be duplicated in the file and in the command.
+
+The outputs will be in the form `source header [locations]`, which means those can be accessed by indexing the output of the `yacc` call:
+
+```meson
+p1 = codegen.yacc('parser.y', header : '@BASENAME@.h', table : '@BASENAME@.tab.h')
+headers = [l1[1], l1[2]]  # [header, table]
```

Comment:
Shouldn't these be `p1[..]`?

Also, I've seen meson use the indexing thing elsewhere, it's a bit weird to me, `p1.header` would be much clearer. It feels like you need a NamedTuple sort of construct in meson for these sorts of things...

Reply:
Ah, yes, those should all be `p1` or `l1`.

The problem with having a `.header` is there's no requirement that a custom target output a header. Or even source code. Or that you want to access all of the headers at once.
```meson
ct = custom_target(
   'ct',
   output : ['code_for_lib.c', 'code_for_exe.c', 'lib_header.h', 'exe_header.h'],
   ...
)

l = shared_library('l', ct[0], ct[2])
e = executable('e', ct[1], ct[2], ct[3], link_with : l)

ct2 = custom_target(
   'generate_tarballs',
   output : ['tarball.tar.gz', 'tarball.tar.bz2', 'tarball.tar.xz'],
   ...
)
```
Generally, you just want to put this all in a `declare_dependency` anyway, since you need to propagate the headers with the library to ensure proper build ordering:
```meson
ct = custom_target(
   'ct',
   output : ['code_for_lib.c', 'code_for_exe.c', 'lib_header.h', 'exe_header.h'],
   ...
)

l = shared_library('l', ct[0], ct[2])
lib = declare_dependency(
  link_with : l,
  sources : ct[2],
  include_directories : '.'
)
```